rskj-core/src/main/java/co/rsk/config/ConfigLoader.java

```diff
@@ -55,14 +51,59 @@ public ConfigLoader(CliArgs<NodeCliOptions, NodeCliFlags> cliArgs) {
         this.cliArgs = Objects.requireNonNull(cliArgs);
     }
 
+    /**
+     * Loads configurations from different sources with the following precedence:
+     * 1. Command line arguments
+     * 2. Environment variables
+     * 3. System properties
+     * 4. User configuration file
+     * 5. Installer configuration file
+     * 6. Default settings per network in resources/[network].conf
+     * 7. Default settings for all networks in resources/reference.conf
+     *
+     * <p>
+     * If the <b><blockchain.config.verify/b> setting is {@code true} (either set in a .conf file or via <b>--verify-config</b> command line flag),
+     * then the loaded config wll be tested against the setting names defined in the expected.conf config file.
+     *
+     * Note:
+     *  1. The <b><blockchain.config.verify/b> setting is {@code false} by default.
+     *  2. Config verification process of matching actual and expected configs is recursive and takes into account appropriate setting names
+     *      and values which are collections of other settings (LIST's and OBJECT's).
+     *
+     * @throws RskConfigurationException on configuration errors
+     *
+     * @see ConfigProblems
+     */
     public Config getConfig() {
-        Config userConfig = getConfigFromCliArgs()
-                .withFallback(ConfigFactory.systemProperties())
-                .withFallback(ConfigFactory.systemEnvironment())
-                .withFallback(getUserCustomConfig())
-                .withFallback(getInstallerConfig());
+        Config cliConfig = getConfigFromCliArgs();
+        Config systemPropsConfig = ConfigFactory.systemProperties();
+        Config systemEnvConfig = ConfigFactory.systemEnvironment();
+        Config userCustomConfig = getUserCustomConfig();
+        Config installerConfig = getInstallerConfig();
+
+        Config userConfig = ConfigFactory.empty()
+                .withFallback(cliConfig)
+                .withFallback(systemPropsConfig)
+                .withFallback(systemEnvConfig)
+                .withFallback(userCustomConfig)
+                .withFallback(installerConfig);
         Config networkBaseConfig = getNetworkDefaultConfig(userConfig);
-        return userConfig.withFallback(networkBaseConfig);
+        Config unifiedConfig = userConfig.withFallback(networkBaseConfig);
+
+        if (unifiedConfig.getBoolean(SystemProperties.PROPERTY_BC_VERIFY)) {
+            Config expectedConfig = ConfigFactory.parseResourcesAnySyntax(EXPECTED_RESOURCE_PATH)
+                    .withFallback(systemPropsConfig)
+                    .withFallback(systemEnvConfig);
+
+            ArrayList<String> problems = new ArrayList<>();
+            verify("", expectedConfig.root(), unifiedConfig.root(), problems);
+            if (!problems.isEmpty()) {
+                throw new RskConfigurationException("Verification of node configs has failed. The following problems were found: "
+                        + String.join("; ", problems));
```

Comment:
I can see this getting overly verbose and quite difficult to read and search for problems, should multiple configuration problems be present. I suggest making `verify` return a boolean instead, signaling if the verification was successful, and logging each of the misconfigurations instead of pushing them into `problems`.

Reply:
I agree and disagree.
**Agree:** I would iterate over problems and log each of them as a separate line.
**Disagree:** I think the method works fine as it is, right now we don't unit test it directly, but it could be directly unit tested, and it's a richer interface to have it return a problems array instead of a boolean. What I would add is an overload:
```
 private static List<String> verify(ConfigValue expectedRoot, ConfigValue actualRoot) {
    ArrayList<String> problems = new ArrayList<>();
    verify("", expectedConfig.root(), unifiedConfig.root(), problems);
    return problems;
}
``` 

IMHO it's nice to add an overload in such cases to make the calling code more oblivious to the internal details of the recursive method.

Reply:
I just don't want to overarchitect this solution. Good point about doing verification all the time and only throwing an exception if the flag is present.